test/tapeimage.cpp

```diff
@@ -323,6 +323,68 @@ TEST_CASE(
     CHECK(err == LFP_OK);
 }
 
+TEST_CASE(
+    "Peel off the current protocol and expose the underlying one"
+    "[tapeimage][peel]") {
+
+    const auto file = std::vector< unsigned char > {
+        0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        0x14, 0x00, 0x00, 0x00,
+
+        /* begin body */
+        0x01, 0x02, 0x03, 0x04,
+        0x05, 0x06, 0x07, 0x08,
+        /* end body */
+
+        0x01, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        0x20, 0x00, 0x00, 0x00,
+
+        0x01, 0x00, 0x00, 0x00,
+        0x14, 0x00, 0x00, 0x00,
+        0x2C, 0x00, 0x00, 0x00,
+    };
+
+    const auto expected1 = std::vector< unsigned char > {
+        0x01, 0x02, 0x03, 0x04,
+    };
+
+    const auto expected2 = std::vector< unsigned char > {
+        0x05, 0x06, 0x07, 0x08,
+    };
+
+    auto* mem = lfp_memfile_openwith(file.data(), file.size());
+    auto* tif = lfp_tapeimage_open(mem);
+
+    auto out = std::vector< unsigned char >(4, 0xFF);
+    std::int64_t bytes_read = -1;
+    auto err = lfp_readinto(tif, out.data(), 4, &bytes_read);
+
+    CHECK(bytes_read == 4);
+    CHECK(err == LFP_OK);
+    CHECK_THAT(out, Equals(expected1));
+
+    std::int64_t tif_tell;
+    lfp_tell(tif, &tif_tell);
+    CHECK(tif_tell == 4);
+
+    lfp_protocol* flesh;
+    err = lfp_peel(tif, &flesh);
+    CHECK(err == LFP_OK);
+
+    std::int64_t mem_tell;
+    lfp_tell(flesh, &mem_tell);
+    CHECK(mem_tell == 16);
+
+    err = lfp_readinto(flesh, out.data(), 4, &bytes_read);
+    CHECK(bytes_read == 4);
+    CHECK(err == LFP_OK);
+    CHECK_THAT(out, Equals(expected2));
+
+    lfp_close(tif);
```

Comment:
Aren't you supposed to close mem here separately as it doesn't belong to tif anymore?

`+` actually having some tests on tif (as a childless protocol) would be cool. It should basically scream to all the operations on it except for close.

Reply:
1) yup
2) I dont see the point of such tests. After peeling a protocol any operation on it should be considered undefined behavior. (should probably clarify that in the docs). Btw, it does not scream as loud as you might expect. E.g. for the `tapeimage::eof()` would not scream, but `rp66::eof()` would. For tell its the opposite: `tapeimage::tell` would scream but `rp66::tell` would still "work"-